base_delivery_carrier_label/models/stock_picking.py

```diff
@@ -0,0 +1,190 @@
+# Copyright 2012-2015 Akretion <http://www.akretion.com>.
+# Copyright 2013-2016 Camptocamp SA
+# License AGPL-3.0 or later (https://www.gnu.org/licenses/agpl).
+
+import logging
+
+from odoo import _, api, fields, models
+from odoo.exceptions import UserError
+
+_logger = logging.getLogger(__name__)
+
+
+class StockPicking(models.Model):
+    _inherit = "stock.picking"
+
+    carrier_id = fields.Many2one(
+        comodel_name="delivery.carrier",
+        string="Carrier",
+        states={"done": [("readonly", True)]},
+    )
+    carrier_code = fields.Char(related="carrier_id.code", readonly=True)
+    option_ids = fields.Many2many(
+        comodel_name="delivery.carrier.option", string="Options"
+    )
+
+    def get_shipping_label_values(self, label):
+        self.ensure_one()
+        return {
+            "name": label["name"],
+            "res_id": self.id,
+            "res_model": "stock.picking",
+            "datas": label["file"],
+            "file_type": label["file_type"],
+        }
+
+    def attach_shipping_label(self, label):
+        """Attach a label returned by generate_shipping_labels to a picking"""
+        self.ensure_one()
+        data = self.get_shipping_label_values(label)
+        if label.get("package_id"):
+            data["package_id"] = label["package_id"]
+            if label.get("tracking_number"):
+                self.env["stock.quant.package"].browse(label["package_id"]).write(
+                    {"parcel_tracking": label.get("tracking_number")}
+                )
+        context_attachment = self.env.context.copy()
+        # remove default_type setted for stock_picking
+        # as it would try to define default value of attachement
+        if "default_type" in context_attachment:
+            del context_attachment["default_type"]
+        return (
+            self.env["shipping.label"].with_context(**context_attachment).create(data)
+        )
+
+    def _set_a_default_package(self):
+        """Pickings using this module must have a package
+        If not this method put it one silently
+        """
+        for picking in self:
+            move_lines = picking.move_line_ids.filtered(
+                lambda s: not s.result_package_id
+            )
+            if move_lines:
+                picking._put_in_pack(move_lines)
+
+    def send_to_shipper(self):
+        self.ensure_one()
+        if self.env.context.get("set_default_package", True):
+            self._set_a_default_package()
```

Comment:
@florian-dacosta @glitchov Could you help me and explain why we need to set a new package ? For what purpose ? Thanks

Reply:
Ok, I see that's because we want to store the tracking on package side. But that should be optional IMHO and let by default the tracking on picking side.

Reply:
@rousseldenis 
I think, the real goal was to ease the carrier modules. 
There were/are sometimes module with quite a lot of code to manage the case where we don't have any package on the picking and we have to generate a label, and the case where we have packages and so we have to generate a label with a pack... 
I guess the idea was kind of : the pack should be mandatory when generating a carrier label, but we don't want to force the user that don't need to use pack in Odoo to fill the pack on each picking

In reality there is always at least a pack, it is just not filled in Odoo for some users. And the carrier webservice usually expects some data relative to a pack.
I know the delivery_roulier is a lot simpler knowing it does not have to manage products outside a pack for generating carrier labels.
But... I guess it can be an issue, mainly in the case we don't want to manage pack in Odoo and we have action after the delivery (return or interwarehouse shipments, etc...)

Reply:
Thanks for your answer. I think I need to dig a little bit in this in order to identify corner cases.